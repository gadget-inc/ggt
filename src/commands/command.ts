import arg from "arg";
import assert from "node:assert";
import type { Context } from "src/services/command/context.js";
import { relativeToThisFile } from "src/services/config/paths.js";
import type { Promisable } from "type-fest";

export type Usage = () => string;

export type Command = (ctx: Context) => Promisable<void>;

export type CommandModule = {
  usage: Usage;
  command: Command;
};

export const AvailableCommands = ["sync", "list", "login", "logout", "whoami", "version", "deploy"] as const;

type AvailableCommand = (typeof AvailableCommands)[number];

export const isAvailableCommand = (value: unknown): value is AvailableCommand => {
  return AvailableCommands.includes(value as AvailableCommand);
};

export const importCommandModule = async (command: AvailableCommand): Promise<CommandModule> => {
  assert(isAvailableCommand(command), `invalid command: ${command}`);
  return (await import(relativeToThisFile(`../../commands/${command}.js`))) as CommandModule;
};

export const rootArgsSpec = {
  "--help": Boolean,
  "-h": "--help",
  "--verbose": arg.COUNT,
  "-v": "--verbose",
  "--json": Boolean,

  // deprecated
  "--debug": "--verbose",
};

export type RootArgs = arg.Result<typeof rootArgsSpec>;
