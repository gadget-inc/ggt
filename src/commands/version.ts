import type { Command, Usage } from "../services/command/command.js";
import { packageJson } from "../services/config/package-json.js";
import { println } from "../services/output/print.js";
import { sprint } from "../services/output/sprint.js";

export const usage: Usage = () => sprint`
  Print this version of ggt.

  {gray Usage}
        ggt version
  
  {gray Updating ggt}
        When there is a new release of ggt, running ggt will show you a message letting you
        know that an update is available.
`;

export const command: Command = (_ctx) => {
  println(packageJson.version);
};
