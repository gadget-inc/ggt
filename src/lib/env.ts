import type { Paths } from "env-paths";
import envPaths from "env-paths";

/**
 * Captures the name and nature of the environment
 */
export class Env {
  static get value(): string {
    return process.env["GGT_ENV"] || "production";
  }

  static get paths(): Paths {
    return envPaths("gadget", { suffix: this.productionLike ? "" : this.value });
  }

  static get developmentLike(): boolean {
    return this.value.startsWith("development");
  }

  static get testLike(): boolean {
    return this.value.startsWith("test");
  }

  static get developmentOrTestLike(): boolean {
    return this.developmentLike || this.testLike;
  }

  static get productionLike(): boolean {
    return this.value.startsWith("production");
  }
}
