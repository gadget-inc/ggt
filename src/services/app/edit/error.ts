import type { GraphQLError } from "graphql";
import assert from "node:assert";
import pluralize from "pluralize";
import type { CloseEvent, ErrorEvent } from "ws";
import { sprint } from "../../output/print.js";
import { CLIError, IsBug } from "../../output/report.js";
import { uniq } from "../../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString } from "../../util/is.js";
import { serializeError } from "../../util/object.js";
import type { GraphQLMutation, GraphQLQuery, GraphQLSubscription } from "./operation.js";

export class EditError extends CLIError {
  isBug = IsBug.MAYBE;

  override cause: string | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly request: GraphQLQuery | GraphQLMutation | GraphQLSubscription,
    cause: unknown,
  ) {
    super("An error occurred while communicating with Gadget");

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override render(): string {
    let body = "";

    switch (true) {
      case isGraphQLErrors(this.cause): {
        const errors = uniq(this.cause.map((x) => x.message));
        body = sprint`
          Gadget responded with the following ${pluralize("error", errors.length, false)}:

            • ${errors.join("\n            • ")}
        `;
        break;
      }
      case isCloseEvent(this.cause):
        body = "The connection to Gadget closed unexpectedly.";
        break;
      case isErrorEvent(this.cause) || isError(this.cause):
        body = this.cause.message;
        break;
      default:
        body = this.cause;
        break;
    }

    return this.message + "\n\n" + body;
  }
}
