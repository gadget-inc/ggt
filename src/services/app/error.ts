import type { GraphQLError } from "graphql";
import assert from "node:assert";
import pluralize from "pluralize";
import type { CloseEvent, ErrorEvent } from "ws";
import { GGTError, IsBug } from "../output/report.js";
import { sprint } from "../output/sprint.js";
import { uniq } from "../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString, isStringArray } from "../util/is.js";
import { serializeError } from "../util/object.js";
import type { GraphQLMutation, GraphQLQuery, GraphQLSubscription } from "./edit/operation.js";

export class ClientError extends GGTError {
  public isBug: IsBug;

  override cause: string | string[] | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly request: GraphQLQuery | GraphQLMutation | GraphQLSubscription,
    cause: unknown,
    isBug?: IsBug,
  ) {
    super("An error occurred while communicating with Gadget");

    if (isBug) {
      this.isBug = isBug;
    } else {
      this.isBug = IsBug.MAYBE;
    }

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isStringArray(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override render(): string {
    let body: string;

    switch (true) {
      case isGraphQLErrors(this.cause): {
        const errors = uniq(this.cause.map((x) => x.message));
        body = sprint`
          Gadget responded with the following ${pluralize("error", errors.length, false)}:

            • ${errors.join("\n            • ")}
        `;
        break;
      }
      case isCloseEvent(this.cause):
        body = "The connection to Gadget closed unexpectedly.";
        break;
      case isErrorEvent(this.cause) || isError(this.cause):
        body = this.cause.message;
        break;
      case isStringArray(this.cause):
        if (this.cause.length === 1) {
          body = String(this.cause[0]);
        } else {
          body = this.cause.join(", ");
        }
        break;
      default:
        body = this.cause;
        break;
    }

    return this.message + "\n\n" + body;
  }
}

export class AuthenticationError extends ClientError {
  constructor(request: GraphQLQuery | GraphQLMutation | GraphQLSubscription) {
    super(request, "Request authentication failed due to the session expiring while running the command. Please sign-in again.", IsBug.NO);
  }
}
