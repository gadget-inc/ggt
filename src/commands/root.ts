import ms from "ms";
import { AvailableCommands, importCommandModule, isAvailableCommand, type Usage } from "../services/command/command.js";
import { Context } from "../services/command/context.js";
import { verbosityToLevel } from "../services/output/log/level.js";
import { createLogger } from "../services/output/log/logger.js";
import { reportErrorAndExit } from "../services/output/report.js";
import { sprint } from "../services/output/sprint.js";
import { warnIfUpdateAvailable } from "../services/output/update.js";
import { sortBySimilar } from "../services/util/collection.js";
import { isNil } from "../services/util/is.js";

const log = createLogger({ name: "root" });

export const usage: Usage = () => sprint`
    The command-line interface for Gadget

    {bold USAGE}
      ggt [COMMAND]

    {bold COMMANDS}
      sync           Sync your Gadget application's source code
      list           List your apps
      login          Log in to your account
      logout         Log out of your account
      whoami         Print the currently logged in account
      version        Print the version of ggt

    {bold FLAGS}
      -h, --help     Print command's usage
      -v, --verbose  Print verbose output
          --json     Print output as JSON

    For more information on a specific command, use 'ggt [COMMAND] --help'
`;

export const command = async (): Promise<void> => {
  const ctx = new Context();

  await warnIfUpdateAvailable();

  if (ctx.args["--json"]) {
    process.env["GGT_LOG_FORMAT"] = "json";
  }

  if (ctx.args["--verbose"]) {
    process.env["GGT_LOG_LEVEL"] = verbosityToLevel(ctx.args["--verbose"]).toString();
  }

  const command = ctx.args._.shift();
  if (isNil(command)) {
    log.println(usage());
    process.exit(0);
  }

  if (!isAvailableCommand(command)) {
    const [closest] = sortBySimilar(command, AvailableCommands);
    log.println`
      Unknown command {yellow ${command}}

      Did you mean {blueBright ${closest}}?

      Run {gray ggt --help} for usage
    `;
    process.exit(1);
  }

  const commandModule = await importCommandModule(command);

  if (ctx.args["--help"]) {
    log.println(commandModule.usage());
    process.exit(0);
  }

  try {
    await commandModule.command(ctx);
  } catch (error) {
    await reportErrorAndExit(error);
  }

  for (const signal of ["SIGINT", "SIGTERM"] as const) {
    process.once(signal, () => {
      log.trace("received signal", { signal });
      log.println` Stopping... {gray Press Ctrl+C again to force}`;
      ctx.abort();

      // when ggt is run via npx, and the user presses ctrl+c, npx
      // sends sigint twice in quick succession. in order to prevent
      // the second sigint from triggering the force exit listener,
      // we wait a bit before registering it
      setTimeout(() => {
        process.once(signal, () => {
          log.println(" Exiting immediately");
          process.exit(1);
        });
      }, ms("100ms")).unref();
    });
  }
};
