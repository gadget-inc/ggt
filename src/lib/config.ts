import fs from "fs-extra";
import path from "path";
import { ignoreEnoent } from "./enoent";
import { Env } from "./env";

type WrittenConfig = { session?: string };

/**
 * Holds the configuration of the CLI.
 */
export class Config {
  /**
   * The session token of the user.
   */
  static session?: string;

  /**
   * The path to the configuration file.
   */
  static get filepath(): string {
    return path.join(Env.paths.config, "config.json");
  }

  /**
   * Saves the configuration to the filesystem.
   */
  static save(this: void): void {
    const config: WrittenConfig = { session: Config.session };
    fs.outputJsonSync(Config.filepath, config, { spaces: 2, mode: 0o600 });
  }

  /**
   * Reloads the configuration from the filesystem.
   */
  static reload(): void {
    try {
      const config = fs.readJSONSync(Config.filepath) as WrittenConfig;
      Config.session = config.session;
    } catch (error) {
      ignoreEnoent(error);
      Config.session = undefined;
    }
  }
}

Config.reload();
