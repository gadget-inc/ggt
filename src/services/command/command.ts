import assert from "node:assert";
import { pathToFileURL } from "node:url";
import type { EmptyObject, Promisable } from "type-fest";
import type { RootArgs } from "../../commands/root.js";
import { config } from "../config/config.js";
import { relativeToThisFile } from "../util/paths.js";
import type { ArgsDefinition } from "./arg.js";
import type { Context } from "./context.js";

/**
 * The list of available commands.
 *
 * 1. Every command corresponds to a file inside of src/commands/
 * 2. The order determines the order of commands in the README
 */
export const Commands = ["dev", "deploy", "status", "push", "pull", "add", "open", "list", "login", "logout", "whoami", "version"] as const;

/**
 * One of the commands in {@link Commands}.
 */
export type AvailableCommand = (typeof Commands)[number];

/**
 * Checks if a string is a valid command.
 *
 * @param command - The string to check
 * @returns Whether the string is a valid command
 */
export const isAvailableCommand = (command: string): command is AvailableCommand => {
  return Commands.includes(command as AvailableCommand);
};

/**
 * A command module is a file in the src/commands/ directory.
 */
export type CommandModule<Args extends ArgsDefinition = EmptyObject, ParentArgs extends ArgsDefinition = RootArgs> = {
  /**
   * The command's {@link ArgsDefinition args}.
   */
  args?: Args;

  /**
   * The command's {@link Usage usage}.
   */
  usage: Usage;

  /**
   * The command's {@link Command command}.
   *
   * TODO: rename this to `run`.
   */
  command: Command<Args, ParentArgs>;
};

/**
 * A {@linkcode Command command}'s usage is a function that returns a
 * string describing how to use the command. The function receives its
 * parent command's context.
 */
export type Usage = (ctx: Context) => string;

/**
 * The function that is run when the command is called.
 *
 * @param ctx - A {@linkcode Context} with the command's {@linkcode Args} and {@linkcode ParentArgs}.
 */
export type Command<Args extends ArgsDefinition = EmptyObject, ParentArgs extends ArgsDefinition = RootArgs> = (
  ctx: Context<Args, ParentArgs>,
) => Promisable<void>;

/**
 * Imports a command module.
 *
 * @param cmd - The command to import
 * @see {@linkcode CommandModule}
 */
export const importCommand = async (cmd: AvailableCommand): Promise<CommandModule> => {
  assert(isAvailableCommand(cmd), `invalid command: ${cmd}`);
  let commandPath = relativeToThisFile(`../../commands/${cmd}.js`);
  if (config.windows) {
    // https://github.com/nodejs/node/issues/31710
    commandPath = pathToFileURL(commandPath).toString();
  }
  return (await import(commandPath)) as CommandModule;
};
