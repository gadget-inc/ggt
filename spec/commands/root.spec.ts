import process from "node:process";
import { inspect } from "node:util";
import { afterEach, assert, beforeEach, describe, expect, it, vi } from "vitest";
import { spyOnImplementing } from "vitest-mock-process";
import { command as root } from "../../src/commands/root.js";
import * as command from "../../src/services/command/command.js";
import { importCommand, type CommandSpec } from "../../src/services/command/command.js";
import { config } from "../../src/services/config/config.js";
import { Level } from "../../src/services/output/log/level.js";
import * as update from "../../src/services/output/update.js";
import { noop, noopThis, type AnyFunction } from "../../src/services/util/function.js";
import { isAbortError } from "../../src/services/util/is.js";
import { PromiseSignal } from "../../src/services/util/promise.js";
import { withEnv } from "../__support__/env.js";
import { expectReportErrorAndExit } from "../__support__/error.js";
import { expectProcessExit } from "../__support__/process.js";
import { expectStdout } from "../__support__/stream.js";
import { mockVersion } from "../__support__/version.js";

describe("root", () => {
  mockVersion();

  beforeEach(() => {
    spyOnImplementing(process, "once", noopThis);

    // don't check for updates
    vi.spyOn(update, "warnIfUpdateAvailable").mockResolvedValue();
  });

  afterEach(() => {
    expect(update.warnIfUpdateAvailable).toHaveBeenCalled();
  });

  it("prints root usage when no command is given", async () => {
    process.argv = ["node", "ggt"];

    await expectProcessExit(root);

    expectStdout().toMatchInlineSnapshot(`
      "The command-line interface for Gadget

      USAGE
        ggt [COMMAND]

      COMMANDS
        sync           Sync your Gadget application's source code
        deploy         Deploy your app to production
        list           List your apps
        login          Log in to your account
        logout         Log out of your account
        whoami         Print the currently logged in account
        version        Print the version of ggt

      FLAGS
        -h, --help     Print command's usage
        -v, --verbose  Print verbose output
            --json     Print output as JSON

      For more information on a specific command, use 'ggt [COMMAND] --help'
      "
    `);
  });

  it("prints out a helpful message when an unknown command is given", async () => {
    process.argv = ["node", "ggt", "foobar"];

    await expectProcessExit(root, 1);

    expectStdout().toMatchInlineSnapshot(`
      "Unknown command foobar

      Did you mean login?

      Run ggt --help for usage
      "
    `);
  });

  it("sets GGT_LOG_FORMAT=json when --json is given", async () => {
    await withEnv({ GGT_LOG_FORMAT: undefined }, async () => {
      expect(config.logFormat).toBe("pretty");

      process.argv = ["node", "ggt", "--json"];
      await expectProcessExit(root);

      expect(process.env["GGT_LOG_FORMAT"]).toBe("json");
      expect(config.logFormat).toBe("json");
    });
  });

  it.each([
    [Level.INFO, "-v"],
    [Level.DEBUG, "-vv"],
    [Level.TRACE, "-vvv"],
  ])("sets GGT_LOG_LEVEL=%d when %s is given", async (level, flag) => {
    await withEnv({ GGT_LOG_LEVEL: undefined }, async () => {
      expect(config.logLevel).toBe(Level.PRINT);

      process.argv = ["node", "ggt", flag];
      await expectProcessExit(root);

      expect(process.env["GGT_LOG_LEVEL"]).toBe(String(level));
      expect(config.logLevel).toBe(level);
    });
  });

  const signals = ["SIGINT", "SIGTERM"] as const;
  it.each(signals)("calls ctx.abort() on %s", async (expectedSignal) => {
    const aborted = new PromiseSignal();

    vi.spyOn(command, "isAvailableCommand").mockReturnValueOnce(true);
    vi.spyOn(command, "importCommand").mockResolvedValueOnce({
      usage: () => "abort test",
      command: (ctx) => {
        ctx.signal.addEventListener("abort", (reason) => {
          assert(isAbortError(reason), `reason isn't an AbortError: ${inspect(reason)}`);
          aborted.resolve();
        });
      },
    });

    let signalled = false;
    let onSignal: AnyFunction;

    spyOnImplementing(process, "once", (actualSignal, cb) => {
      signalled ||= actualSignal === expectedSignal;
      expect(signals).toContain(actualSignal);
      onSignal = cb;
      return process;
    });

    process.argv = ["node", "ggt", "test"];
    await root();

    expect(signalled).toBe(true);
    onSignal!();

    await aborted;
  });

  describe.each(command.AvailableCommands)("when %s is given", (name) => {
    let cmd: CommandSpec;

    beforeEach(async () => {
      cmd = await importCommand(name);
      vi.spyOn(cmd, "command").mockImplementation(noop);
    });

    it.each(["--help", "-h"])("prints the usage when %s is passed", async (flag) => {
      process.argv = ["node", "ggt", name, flag];

      await expectProcessExit(root);

      expectStdout().toEqual(cmd.usage() + "\n");
    });

    it("runs the command", async () => {
      process.argv = ["node", "ggt", name];

      await root();

      expect(cmd.command).toHaveBeenCalled();
    });

    it("reports and exits if an error occurs", async () => {
      const error = new Error("boom!");
      vi.spyOn(cmd, "command").mockRejectedValueOnce(error);

      process.argv = ["node", "ggt", name];

      void root();
      await expectReportErrorAndExit(error);

      expect(cmd.command).toHaveBeenCalled();
    });
  });
});
