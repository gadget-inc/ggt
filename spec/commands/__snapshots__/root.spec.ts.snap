// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`root > when deploy is given > prints the usage when --help is passed 1`] = `
"Deploy your Gadget application's development source code to production.

USAGE
  ggt deploy [DIRECTORY] [--app=<name>]

ARGUMENTS
  DIRECTORY         The directory to sync files to and deploy (default: \\".\\")

FLAGS
  -a, --app=<name>  The Gadget application to deploy
      --force       Deploy the Gadget application regardless of any issues it may have

DESCRIPTION
  Deploy allows you to deploy your current Gadget application in development to production.

  It detects if local files are up to date with remote and if the Gadget application
  is in a deployable state. If there are any issues, it will display them and ask if
  you would like to deploy anyways.

  Note:
    • If local files are not up to date or have not recently been synced with remote ones,
      you will be prompted to run a one-time sync to ensure the files remain consistent with
      what is on the remote.
    • You may wish to keep ggt sync running in the background before trying to run ggt deploy

EXAMPLE
  $ ggt deploy ~/gadget/example --app example

  App         example
  Editor      https://example.gadget.app/edit
  Playground  https://example.gadget.app/api/graphql/playground
  Docs        https://docs.gadget.dev/api/example

  Endpoints
    • https://example.gadget.app
    • https://example--development.gadget.app


  Building frontend assets ...
  ✔ DONE

  Setting up database ...
  ✔ DONE

  Copying development ...
  ✔ DONE

  Restarting app ...
  ✔ DONE

  Deploy completed. Good bye!
"
`;

exports[`root > when deploy is given > prints the usage when -h is passed 1`] = `
"Deploy your Gadget application's development source code to production.

USAGE
  ggt deploy [DIRECTORY] [--app=<name>]

ARGUMENTS
  DIRECTORY         The directory to sync files to and deploy (default: \\".\\")

FLAGS
  -a, --app=<name>  The Gadget application to deploy
      --force       Deploy the Gadget application regardless of any issues it may have

DESCRIPTION
  Deploy allows you to deploy your current Gadget application in development to production.

  It detects if local files are up to date with remote and if the Gadget application
  is in a deployable state. If there are any issues, it will display them and ask if
  you would like to deploy anyways.

  Note:
    • If local files are not up to date or have not recently been synced with remote ones,
      you will be prompted to run a one-time sync to ensure the files remain consistent with
      what is on the remote.
    • You may wish to keep ggt sync running in the background before trying to run ggt deploy

EXAMPLE
  $ ggt deploy ~/gadget/example --app example

  App         example
  Editor      https://example.gadget.app/edit
  Playground  https://example.gadget.app/api/graphql/playground
  Docs        https://docs.gadget.dev/api/example

  Endpoints
    • https://example.gadget.app
    • https://example--development.gadget.app


  Building frontend assets ...
  ✔ DONE

  Setting up database ...
  ✔ DONE

  Copying development ...
  ✔ DONE

  Restarting app ...
  ✔ DONE

  Deploy completed. Good bye!
"
`;

exports[`root > when list is given > prints the usage when --help is passed 1`] = `
"List the apps available to the currently logged in user.

USAGE
  ggt list

EXAMPLE
  $ ggt list
    Slug    Domain
    ─────── ──────────────────
    my-app  my-app.gadget.app
    example example.gadget.app
    test    test.gadget.app
"
`;

exports[`root > when list is given > prints the usage when -h is passed 1`] = `
"List the apps available to the currently logged in user.

USAGE
  ggt list

EXAMPLE
  $ ggt list
    Slug    Domain
    ─────── ──────────────────
    my-app  my-app.gadget.app
    example example.gadget.app
    test    test.gadget.app
"
`;

exports[`root > when login is given > prints the usage when --help is passed 1`] = `
"Log in to your account.

USAGE
  ggt login

EXAMPLE
  $ ggt login
    We've opened Gadget's login page using your default browser.

    Please log in and then return to this terminal.

    Hello, Jane Doe (jane@example.com)
"
`;

exports[`root > when login is given > prints the usage when -h is passed 1`] = `
"Log in to your account.

USAGE
  ggt login

EXAMPLE
  $ ggt login
    We've opened Gadget's login page using your default browser.

    Please log in and then return to this terminal.

    Hello, Jane Doe (jane@example.com)
"
`;

exports[`root > when logout is given > prints the usage when --help is passed 1`] = `
"Log out of your account.

USAGE
  ggt logout

EXAMPLE
  $ ggt logout
    Goodbye
"
`;

exports[`root > when logout is given > prints the usage when -h is passed 1`] = `
"Log out of your account.

USAGE
  ggt logout

EXAMPLE
  $ ggt logout
    Goodbye
"
`;

exports[`root > when sync is given > prints the usage when --help is passed 1`] = `
"Sync your Gadget environment's source code with your local filesystem.

USAGE
  ggt sync [DIRECTORY]

ARGUMENTS
  DIRECTORY                  The directory to sync files to (default: \\".\\")

FLAGS
  -a, --app=<name>           The Gadget application to sync files to
      --prefer=<filesystem>  Prefer \\"local\\" or \\"gadget\\" conflicting changes
      --once                 Sync once and exit
      --force                Sync regardless of local filesystem state

DESCRIPTION
  Sync allows you to synchronize your Gadget application's source
  code with your local filesystem.

  While ggt sync is running, local file changes are immediately
  reflected within Gadget, while files that are changed in Gadget are
  immediately saved to your local filesystem.

  Ideal for:
    • Local development with editors like VSCode
    • Storing source code in a Git repository like GitHub

  Sync looks for a \\".ignore\\" file to exclude certain files/directories
  from being synced. The format is identical to Git's.

  These files are always ignored:
    • .DS_Store
    • .gadget
    • .git
    • node_modules

  Note:
    • Sync only works with your development environment
    • Avoid deleting/moving all your files while sync is running
    • Gadget only supports Yarn v1 for dependency installation

EXAMPLE
  $ ggt sync ~/gadget/example --app example

    App         example
    Editor      https://example.gadget.app/edit
    Playground  https://example.gadget.app/api/graphql/playground
    Docs        https://docs.gadget.dev/api/example

    Endpoints
      • https://example.gadget.app
      • https://example--development.gadget.app

    Watching for file changes... Press Ctrl+C to stop

    → Sent 09:06:25 AM
    routes/GET-hello.js  + created

    → Sent 09:06:49 AM
    routes/GET-hello.js  ± updated

    ← Received 09:06:54 AM
    routes/GET-hello.js  ± updated

    ← Received 09:06:56 AM
    routes/GET-hello.js  - deleted
    ^C Stopping... press Ctrl+C again to force

    Goodbye!
"
`;

exports[`root > when sync is given > prints the usage when -h is passed 1`] = `
"Sync your Gadget environment's source code with your local filesystem.

USAGE
  ggt sync [DIRECTORY]

ARGUMENTS
  DIRECTORY                  The directory to sync files to (default: \\".\\")

FLAGS
  -a, --app=<name>           The Gadget application to sync files to
      --prefer=<filesystem>  Prefer \\"local\\" or \\"gadget\\" conflicting changes
      --once                 Sync once and exit
      --force                Sync regardless of local filesystem state

DESCRIPTION
  Sync allows you to synchronize your Gadget application's source
  code with your local filesystem.

  While ggt sync is running, local file changes are immediately
  reflected within Gadget, while files that are changed in Gadget are
  immediately saved to your local filesystem.

  Ideal for:
    • Local development with editors like VSCode
    • Storing source code in a Git repository like GitHub

  Sync looks for a \\".ignore\\" file to exclude certain files/directories
  from being synced. The format is identical to Git's.

  These files are always ignored:
    • .DS_Store
    • .gadget
    • .git
    • node_modules

  Note:
    • Sync only works with your development environment
    • Avoid deleting/moving all your files while sync is running
    • Gadget only supports Yarn v1 for dependency installation

EXAMPLE
  $ ggt sync ~/gadget/example --app example

    App         example
    Editor      https://example.gadget.app/edit
    Playground  https://example.gadget.app/api/graphql/playground
    Docs        https://docs.gadget.dev/api/example

    Endpoints
      • https://example.gadget.app
      • https://example--development.gadget.app

    Watching for file changes... Press Ctrl+C to stop

    → Sent 09:06:25 AM
    routes/GET-hello.js  + created

    → Sent 09:06:49 AM
    routes/GET-hello.js  ± updated

    ← Received 09:06:54 AM
    routes/GET-hello.js  ± updated

    ← Received 09:06:56 AM
    routes/GET-hello.js  - deleted
    ^C Stopping... press Ctrl+C again to force

    Goodbye!
"
`;

exports[`root > when version is given > prints the usage when --help is passed 1`] = `
"Print the version of ggt

USAGE
  ggt version

EXAMPLE
  $ ggt version
    1.2.3
"
`;

exports[`root > when version is given > prints the usage when -h is passed 1`] = `
"Print the version of ggt

USAGE
  ggt version

EXAMPLE
  $ ggt version
    1.2.3
"
`;

exports[`root > when whoami is given > prints the usage when --help is passed 1`] = `
"Show the name and email address of the currently logged in user

USAGE
  ggt whoami

EXAMPLE
  $ ggt whoami
    You are logged in as Jane Doe (jane@example.com)
"
`;

exports[`root > when whoami is given > prints the usage when -h is passed 1`] = `
"Show the name and email address of the currently logged in user

USAGE
  ggt whoami

EXAMPLE
  $ ggt whoami
    You are logged in as Jane Doe (jane@example.com)
"
`;
